Fix composites failing when inputs are different chunk sizes

djhoese opened the pull request:
While working on using the CREFL rayleigh correction on MODIS data I noticed a couple issues. Any composites that use map_blocks or blockwise will fail if they have different chunk sizes for their inputs. This is happening for MODIS data due to the angle datasets being returned as 64-bit floats when they only need to be 32-bit floats (that'll be another PR) and the EWA resampling dynamically determining output chunk size (dask's `auto` chunk size) which takes `dtype` into account. Anyway, I've added a unify_chunks method that is automatically run as part of match_data_arrays that all Composites typically follow.

Replaces #1812 

 - [x] Tests added <!-- for all bug fixes or enhancements -->

djhoese commented:
I'm being plagued by CI HTTP errors.

sfinkens reviewed (commented):
You could split `unify_chunks` into two methods: One for finding the array pairs and one for unifying the chunks. Otherwise LGTM.

djhoese commented:
I decided to split the array pairs logic out to a separate method, but keep the 3 lines for the unify chunks logic. I don't think clean code is supposed to mean a method that only calls other methods and has no other logic.

mraspaud reviewed (commented):
Looks good, just a few comments. Nice work with the compact unify_chunks function.

djhoese commented on satpy/composites/__init__.py:
```diff
@@ -76,8 +73,7 @@ def check_times(projectables):
             raise IncompatibleTimes
-        else:
-            mid_time = (np.max(times) - np.min(times)) / 2 + np.min(times)
+        mid_time = (np.max(times) - np.min(times)) / 2 + np.min(times)
```
This line is just an indentation change. I'd prefer to leave any changes to this logic to another PR. This change was purely to make CodeFactor happy.

I think `times` is any `times` coordinate variable.

mraspaud replied:
yes, I gathered this was just cosmetic, but it got me thinking, sorry :) If times can be large, it might be worth to cache the min and max... otherwise, just ignore this.

djhoese commented on satpy/composites/__init__.py:
```diff
@@ -159,7 +155,33 @@ def apply_modifier_info(self, origin, destination):
+        new_arrays = self.drop_coordinates(data_arrays)
+        return self.unify_chunks(new_arrays)
+
+    def unify_chunks(self, data_arrays):
```
Yes! I didn't know that existed!

djhoese commented on satpy/composites/__init__.py:
```diff
@@ -159,7 +155,33 @@ def apply_modifier_info(self, origin, destination):
+    @staticmethod
+    def _data_arr_to_dask_array_dim_pairs(data_arrays):
+        all_dims = set.union(*(set(x.dims) for x in data_arrays))
+        dim_map = dict(zip(all_dims, string.ascii_lowercase))
```
Yes. I can add that but given that you just pointed out xarray's `unify_chunks` this method is likely not needed.

djhoese commented:
I switched to xarray's `unify_chunks` which removes most of the code that was originally in this PR. Xarray's `unify_chunks` doesn't do this rechunking inplace so it brought to light another issue in the crefl modifier which is now fixed (tests weren't changed so still TDD).

mraspaud reviewed (approved):
LGTM!
